src/pal/src/numa/numa.cpp

```diff
@@ -527,6 +554,103 @@ SetThreadGroupAffinity(
 
 /*++
 Function:
+  SetThreadAffinityMask
+
+See MSDN doc.
+--*/
+DWORD_PTR
+PALAPI
+SetThreadAffinityMask(
+  IN HANDLE hThread,
+  IN DWORD_PTR dwThreadAffinityMask
+)
+{
+    PERF_ENTRY(SetThreadAffinityMask);
+    ENTRY("SetThreadAffinityMask(hThread=%p, dwThreadAffinityMask=%p)\n", hThread, dwThreadAffinityMask);
+
+    CPalThread *pCurrentThread = InternalGetCurrentThread();
+    CPalThread *pTargetThread = NULL;
+    IPalObject *pTargetThreadObject = NULL;
+
+    PAL_ERROR palErr =
+        InternalGetThreadDataFromHandle(pCurrentThread, hThread,
+                                        0, // THREAD_SET_CONTEXT
+                                        &pTargetThread, &pTargetThreadObject);
+
+    if (NO_ERROR != palErr)
+    {
+        ERROR("Unable to obtain thread data for handle %p (error %x)!\n", hThread,
+              palErr);
+        return 0;
+    }
+
+    pthread_t thread = pTargetThread->GetPThreadSelf();
+
+#if HAVE_PTHREAD_GETAFFINITY_NP
+    cpu_set_t prevCpuSet;
+    CPU_ZERO(&prevCpuSet);
+    KAFFINITY prevMask = 0;
+
+    int st = pthread_getaffinity_np(thread, sizeof(cpu_set_t), &prevCpuSet);
+
+    if (st == 0)
+    {
+        for (int i = 0; i < g_possibleCpuCount; i++)
+        {
+            if (CPU_ISSET(i, &prevCpuSet))
+            {
+                prevMask |= ((KAFFINITY)1) << i;
```

Comment:
C++ standard says;
> If the value of the right operand is negative or is greater or equal to the number of bits in the promoted left operand, the behavior is undefined.

You could use min(8 * sizeof(KAFFINITY), g_possibleCpuCount) as the loop limit.